plugins/field-bitmap/src/field-bitmap.ts

```diff
@@ -0,0 +1,595 @@
+      for (const cell of row) {
+        if (cell !== 0 && cell !== 1) {
+          return null;
+        }
+      }
+    }
+    return newValue;
+  }
+
+  /**
+   * Called when a new value has been validated and is about to be set.
+   *
+   * @param newValue The value that's about to be set.
+   */
+  protected override doValueUpdate_(newValue: number[][]) {
+    super.doValueUpdate_(newValue);
+    if (newValue) {
+      this.imgHeight = newValue.length;
+      this.imgWidth = newValue[0] ? newValue[0].length : 0;
+    }
+  }
+
+  /**
+   * Show the bitmap editor dialog.
+   *
+   * @param e Optional mouse event that triggered the field to open, or
+   *    undefined if triggered programmatically.
+   */
+  protected override showEditor_(e?: Event) {
+    const editor = this.dropdownCreate_();
+    Blockly.DropDownDiv.getContentDiv().appendChild(editor);
+    Blockly.DropDownDiv.showPositionedByField(
+      this,
+      this.dropdownDispose.bind(this),
+    );
+  }
+
+  /**
+   * Updates the block display and editor dropdown when the field re-renders.
+   */
+  protected override render_() {
+    super.render_();
+
+    if (!this.getValue()) {
+      return;
+    }
+
+    if (this.blockDisplayPixels) {
+      this.forAllCells((r, c) => {
+        const pixel = this.getPixel(r, c);
+
+        if (this.blockDisplayPixels) {
+          this.blockDisplayPixels[r][c].style.fill = pixel
+            ? FILLED_PIXEL_COLOR
+            : EMPTY_PIXEL_COLOR;
+        }
+        if (this.editorPixels) {
+          this.editorPixels[r][c].style.background = pixel
+            ? FILLED_PIXEL_COLOR
+            : EMPTY_PIXEL_COLOR;
+        }
+      });
+    }
+  }
+
+  /**
+   * Determines whether the field is editable.
+   *
+   * @returns True since it is always editable.
+   */
+  override updateEditable() {
+    const editable = super.updateEditable();
+    // Blockly.Field's implementation sets these classes as appropriate, but
+    // since this field has no text they just mess up the rendering of the grid
+    // lines.
+    const svgRoot = this.getSvgRoot();
+    if (svgRoot) {
+      Blockly.utils.dom.removeClass(svgRoot, 'blocklyNonEditableText');
+      Blockly.utils.dom.removeClass(svgRoot, 'blocklyEditableText');
+    }
+    return editable;
+  }
+
+  /**
+   * Gets the rectangle built out of dimensions matching SVG's <g> element.
+   *
+   * @returns The newly created rectangle of same size as the SVG element.
+   */
+  override getScaledBBox() {
+    const boundingBox = this.getSvgRoot()?.getBoundingClientRect();
+    if (!boundingBox) {
+      throw new Error('Tried to retrieve a bounding box without a rect');
+    }
+    return new Blockly.utils.Rect(
+      boundingBox.top,
+      boundingBox.bottom,
+      boundingBox.left,
+      boundingBox.right,
+    );
+  }
+
+  /**
+   * Creates the bitmap editor and add event listeners.
+   *
+   * @returns The newly created dropdown menu.
+   */
+  private dropdownCreate_() {
```

Comment:
do you also want to rename this one to remove the underscore?

Reply:
Good catch, thanks! Renamed accordingly.